### PR TITLE
BUG: qcut can fail for highly discontinuous data distributions

### LINKED ISSUE #15069: qcut can fail for highly discontinuous data distributions
#### Code Sample, a copy-pastable example if possible

This code fails for any `K`:

```python
# Your code here
K = 100

pd.qcut([0] * K + [1] * (K + 1), 2)
```
#### Problem description

With pandas 0.19.2, I have:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-782385490865> in <module>()
----> 1 pd.qcut([0] * K + [1] * (K + 1), 2)

pandas/tools/tile.py in qcut(x, q, labels, retbins, precision)
    173     bins = algos.quantile(x, quantiles)
    174     return _bins_to_cuts(x, bins, labels=labels, retbins=retbins,
--> 175                          precision=precision, include_lowest=True)
    176 
    177 

pandas/tools/tile.py in _bins_to_cuts(x, bins, right, labels, retbins, precision, name, include_lowest)
    192 
    193     if len(algos.unique(bins)) < len(bins):
--> 194         raise ValueError('Bin edges must be unique: %s' % repr(bins))
    195 
    196     if include_lowest:

ValueError: Bin edges must be unique: array([0, 1, 1])
```
#### Expected Output

We need some kind of option to decide how to assign values to a quantile bucket in the event that two quantiles have the same value prior to the `searchsorted` call. In this case, the appropriate behavior may be to assign all `1` values to the 50% quantile bucket.

### PR DESCRIPTION
**Pull request by puneet29**
Fixes #15069. Needs refactoring

- [x] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
Thanks @puneet29 

However, before this PR can be accepted, you need to write tests / make sure you don't break existing ones - see [contributing to the code base](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#contributing-to-the-code-base):

> pandas is serious about testing and strongly encourages contributors to embrace test-driven development (TDD). This development process “relies on the repetition of a very short development cycle: first the developer writes an (initially failing) automated test case that defines a desired improvement or new function, then produces the minimum amount of code to pass that test.” So, before actually writing any code, you should write your tests. Often the test can be taken from the original GitHub issue. However, it is always worth considering additional use cases and writing corresponding tests.
> 
> Adding tests is one of the most common requests after code is pushed to pandas. Therefore, it is worth getting in the habit of writing tests ahead of time so this is never an issue.

**Comment by puneet29** [comment]
Hey @MarcoGorelli ! Kindly guide me on what to do with the duplicates parameter, because this PR deals with duplicate values in bins. Can you review the code? Thanks a lot.

File: pandas/core/reshape/tile.py [hunk 1]
```diff
@@ -345,6 +345,18 @@ def qcut(
     else:
         quantiles = q
     bins = algos.quantile(x, quantiles)
+
+    # Added code
+    # ------------------------------
```
Comment (jbrockmendel) [inline pandas/core/reshape/tile.py]: pls remove this and the dashes below

File: pandas/core/reshape/tile.py [hunk 2]
```diff
@@ -345,6 +345,18 @@ def qcut(
...
+    unique_bins = set(bins)
+
+    for i, b in enumerate(bins):
+        if(b in unique_bins):
```
Comment (jbrockmendel) [inline pandas/core/reshape/tile.py]: space after "if"

**Comment by jbrockmendel** [comment]
@puneet29 this needs tests so we know that the bug is fixed.  it will probably go in pandas.tests.reshape.test_qcut

**Comment by WillAyd** [comment]
I think this is stale so closing for now, but ping @puneet29 if you'd like to pick back up

**Comment by puneet29** [comment]
Hi, I'm sorry. I will take it up again. I was quite busy before. 😅

**Comment by WillAyd** [comment]
No worries - thanks!

**Comment by MarcoGorelli** [comment]
> Kindly guide me on what to do with the duplicates parameter

From what I've understood reading through the original issue, the output should not change if `duplicates` is 'drop' - it's in the case when the `duplicates` keyword hasn't been set that the expected output should change.

Anyway, if you write a test case, then it'll be easier for anyone reviewing to tell how the output has changed.

**Comment by jreback** [comment]
closing as stale if you want to continue, please open a new PR.

**Pull request closed without merging**